src/components/OpenFileDialog/OpenFileDialog.js

```diff
@@ -0,0 +1,444 @@
+    Button,
+} from '@dhis2/ui'
+import isEqual from 'lodash/isEqual'
+import PropTypes from 'prop-types'
+import React, { useEffect, useMemo, useState } from 'react'
+import { FileList } from './FileList'
+import { NameFilter } from './NameFilter'
+import { styles } from './OpenFileDialog.styles'
+import { OwnerFilter } from './OwnerFilter'
+import { PaginationControls } from './PaginationControls'
+import { VisTypeFilter } from './VisTypeFilter'
+
+const getResourceFromType = type => `${type}s`
+
+const getQuery = type => ({
+    resource: getResourceFromType(type),
+    params: ({
+        sortField = 'name',
+        sortDirection = 'asc',
+        page = 1,
+        filters,
+    }) => {
+        const queryParams = {
+            filter: filters,
+            fields: `id,type,displayName,title,displayDescription,created,lastUpdated,user,access,href`,
+            paging: true,
+            pageSize: 8,
+            page,
+        }
+
+        if (sortDirection !== 'default') {
+            queryParams.order = `${sortField}:${sortDirection}`
+        }
+
+        return queryParams
+    },
+})
+
+export const OpenFileDialog = ({
+    type,
+    open,
+    onClose,
+    onFileSelect,
+    onNew,
+    currentUser,
+}) => {
+    const filesQuery = useMemo(() => getQuery(type), [])
+    const defaultFilters = {
+        searchTerm: '',
+        createdBy: 'all',
+        visType: 'all',
+    }
+
+    const [{ sortField, sortDirection }, setSorting] = useState({
+        sortField: 'displayName',
+        sortDirection: 'asc',
+    })
+    const [page, setPage] = useState(1)
+    const [searchTerm, setSearchTerm] = useState(defaultFilters.searchTerm)
+    const [searchTimeout, setSearchTimeout] = useState(null)
+    const [filters, setFilters] = useState(defaultFilters)
+
+    const formatFilters = ({ searchTerm, createdBy, visType }) => {
+        const queryFilters = []
+
+        switch (createdBy) {
+            case 'byothers':
+                queryFilters.push(`user.id:!eq:${currentUser.id}`)
+                break
+            case 'byme':
+                queryFilters.push(`user.id:eq:${currentUser.id}`)
+                break
+            case 'all':
+            default:
+                break
+        }
+
+        switch (visType) {
+            case 'all':
+                break
+            case 'chart':
+                queryFilters.push('type:!eq:PIVOT_TABLE')
+                break
+            default:
+                queryFilters.push(`type:eq:${visType}`)
+                break
+        }
+
+        if (searchTerm) {
+            queryFilters.push(`name:ilike:${searchTerm}`)
+        }
+
+        return queryFilters
+    }
+
+    const { loading, error, data, refetch } = useDataQuery(
+        { files: filesQuery },
+        { lazy: true }
+    )
+
+    const resetFilters = () => {
+        setFilters(defaultFilters)
+        setSearchTerm(defaultFilters.searchTerm)
+    }
+
+    const getString = (type, key) => {
```

Comment:
Hm, I see what you're doing here, the texts are similar but not exactly the same, so you want to parse them somehow. 
It looks to me like this can be achieved by using interpolation in i18n instead (e.g. `errorTitle` can be replaced by `i18n.t("Couldn't load {{itemType}}", {itemType: type})`) and just switch or use a ternary on the spot if they differ too much (like the `noDataText` that has a completely different phrasing between the cases). 
Your solution is good because its very scalable, but it seems a bit too complex for just changing a single word?

Reply:
As discussed, I moved the function in `utils.js` but kept the current approach until we have the 3rd app that will use this and will need its own strings.